### ISSUE TITLE
NotificationExtenderService holds old pushnotification Data using onesignal notification

### ISSUE THREAD

**Issue by sachinGAcharya** [OneSignal/OneSignal-Android-SDK#302]
Hi I am using onesignal notification for amazon fire tab.I need silent push notification so by refreing [onesignal documentation](https://documentation.onesignal.com/v3.0/docs/android-customizations)

I am using

```Java
public class NotificationExtenderExample extends NotificationExtenderService {

  @Override
  protected boolean onNotificationProcessing(OSNotificationReceivedResult receivedResult) {
    OverrideSettings overrideSettings = new OverrideSettings();
    
    overrideSettings.extender = new NotificationCompat.Extender() {
      @Override
      public NotificationCompat.Builder extend(NotificationCompat.Builder builder) {
        // Sets the background notification color to Green on Android 5.0+ devices.
        return builder.setColor(new BigInteger("FF00FF00", 16).intValue());
      }
    };
    
    OSNotificationDisplayedResult displayedResult = displayNotification(overrideSettings);   
    
    try {
      Log.d("OneSignalExample", "Notification displayed with id: " + displayedResult.androidNotificationId+"  "+receivedResult.payload.additionalData.get("use_case"));
    } catch (JSONException e) {
      e.printStackTrace();
    }
    
    return true; 
  }
}
```

It is giving me silent notification but It is holding all the notification data.Whenever I launch the app I see It has all the previous notification data.Please tell me how to remove storing of notification data.
-- | -- | --

**Comment by jkasten2** [comment]
@sachinGAcharya The SDK holds on to previous notifications for up to 1 week to prevent duplicates, restore notifications, and calculate badge counts.

For silent notifications that are not displayed we could just store their id so the SDK can prevent any possible duplicate messages being received to save space. Is this your main concern of storing previous notifications?

**Comment by sachinGAcharya** [comment]
Thank you @jkasten2 for replaying.My problem is the above class will be called no of times the total notification received till that till.During every time i launch the app.Means If I have rceived 100 notification the above class will be triggered 100 times during launch of app.So I need to prevent that and  the class  should trigger only whenever the notification comes.Please help me how can i acheive this.

**Comment by jkasten2** [comment]
@sachinGAcharya You need to remove `displayNotification` from your code above if you want your notification to be a silent background data only notification. This should prevent it from being called again for the same notification when your app is cold started as it will be counted as proceed then.

**Comment by yasharpm** [comment]
I have faced the same issue. I am handling the received push in background by returning `true` in the extender method. And I don't create any notifications on my own or by calling `displayNotification`.

But on every launch of my app - probably `Application.onCreate()` - the extender service method `onNotificationProcessing` gets called for the notification I received 24 hours ago. We have not tried sending multiple pushes yet. I'm afraid the case that happened for @sachinGAcharya might be my case as well.

The push actually triggers a "wipe data" on the local database. So having data wiped on every launch of the app after a single push, would not be a very pleasant experience for our users!

**Comment by yasharpm** [comment]
I changed my OneSignal gradle version to '3.6.2' and it seems that the problem does not exist on this version.

**Comment by jkasten2** [comment]
@yasharpm Thanks for the details, what version were you on before reverting to `3.6.2`?

**Comment by yasharpm** [comment]
@jkasten2 The version was set as instructed in the documentation `[3.6.2-3.99.99]`. So I suppose it should at least exist on the latest version.

If you need further details I am willing to put time to help.

**Comment by jkasten2** [comment]
Thanks, we will test this out and get back to you.

**Comment by jkasten2** [comment]
@yasharpm `onNotificationProcessing` will fire in two cases;
1. When the notification is first received
2. Each time the notification needs to be restored back to the shade.
   2a. If you prevented the notification from displaying this event should not fire
   2b. Restoring happens when your app is updated, device is rebooted, or the app is cold started

However there is an edge case bug in 2a in the SDK. If the notification was displayed once then your code tries to omit displaying the notification when it is restored it won't be marked as consumed. The `onNotificationProcessing` will continuing to fire for the 'silent' notification on every 'restore' event.

You most likely ran into this issue when you changed your code in your `onNotificationProcessing` after you sent a few notifications that were displayed on the device.

For reference we just fixed this bug in commit b510e7d3cba701f1ef631239d1c7e640d989bed7. It will be included in the next SDK update.
